### DISCUSSION TITLE
How do I correctly use `setValue` in a generic setting?

### DISCUSSION THREAD

**Discussion by JohannesKohlmann** [react-hook-form/react-hook-form#11777] [Q&A]
Hi :)

I am trying to wrap the `setValue` function exposed by `useFormContext` to only set a value under certain conditions, but can't seem to get the typing correct.

[Sandbox](https://codesandbox.io/p/sandbox/fervent-ully-qxdr7g?file=%2Fsrc%2FApp.tsx%3A20%2C3&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cluwe8uc800062e69kqsff9b4%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cluwe8uc800022e69bsot42e9%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cluwe8uc800032e69scx41l7k%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cluwe8uc800052e69rppl8xna%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cluwe8uc800022e69bsot42e9%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluwe8uc800012e69f820n7n1%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cluwegnzi00022e69rsm0qo0a%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A20%252C%2522startColumn%2522%253A3%252C%2522endLineNumber%2522%253A20%252C%2522endColumn%2522%253A3%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cluwe8uc800022e69bsot42e9%2522%252C%2522activeTabId%2522%253A%2522cluwegnzi00022e69rsm0qo0a%2522%257D%252C%2522cluwe8uc800052e69rppl8xna%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluwe8uc800042e69d7lirqtn%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cluwe8uc800052e69rppl8xna%2522%252C%2522activeTabId%2522%253A%2522cluwe8uc800042e69d7lirqtn%2522%257D%252C%2522cluwe8uc800032e69scx41l7k%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cluwe8uc800032e69scx41l7k%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

My use case looks something like this:

```typescript
type FormFields = {
  field1: number;
  field2: string | undefined
};

const SomeComponent = () => {
  const { setValue } = useFormContext<FormFields>();
  
  const setValueIfDefined = <K extends keyof FormFields>(key: K, value: FormFields[K] | undefined) => {
    if(!value) return;
    setValue(key, value);
  }
  setValueIfDefined('field1', 7);
  return null;
}

```

I am however unable to call `setValue` like this, due to:

TS2345: Argument of type `NonNullable<FormFields[K]>` is not assignable to parameter of type
```
K extends `${infer K}.${infer R}` ? K extends keyof FormFields ? R extends Path<FormFields[K]> ? PathValue<...> : never : K extends `${number}` ? never : never : K extends keyof FormFields ? FormFields[K] : K extends `${number}` ? never : never
```

I am using TS 4.9.5 and react-hook-form 7.43.3

Any insight would be greatly appreciated!

**Comment by davidlhw** [comment]
i'm also facing a similar issue and came across https://github.com/react-hook-form/react-hook-form/issues/4694 from 2021 that doesn't seem to help much

the easiest workaround seems to just be to `ts-ignore` it, but of course that's less than ideal
